### PR TITLE
feat(pageserver): add metrics for aux file size

### PR DESCRIPTION
**Pull request by skyzh**
## Problem

ref https://github.com/neondatabase/neon/issues/7443

## Summary of changes

This pull request adds a size estimator for aux files. Each timeline stores a cached `isize` for the estimated total size of aux files. It gets reset on basebackup, and gets updated for each aux file modification. TODO: print a warning when it exceeds the size.

The size metrics is not accurate. Race between `on_basebackup` and other functions could create a negative basebackup size, but the chance is rare. Anyways, this does not impose any extra I/Os to the storage as everything is computed in-memory.

The aux files are only stored on shard 0. As basebackups are only generated on shard 0, only shard 0 will report this metrics.

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: pageserver/src/aux_file.rs [hunk 1]
```diff
@@ -140,6 +143,55 @@ pub fn encode_file_value(files: &[(&str, &[u8])]) -> anyhow::Result<Vec<u8>> {
     Ok(encoded)
 }
 
+/// An estimation of the size of aux files.
+pub struct AuxFileSizeEstimator {
+    aux_file_size_gauge: IntGauge,
+    size: Arc<std::sync::Mutex<Option<isize>>>,
```
Comment (arpad-m) [inline pageserver/src/aux_file.rs:149]: can this be an atomic?

Comment (skyzh) [reply pageserver/src/aux_file.rs:149]: We need something like `Atomic<Option<isize>>`, which is only possible with manually-implemented compare-with-swap. As the patch is not on the critical path, I prefer keep it simple.

File: pageserver/src/context.rs [hunk 2]
```diff
@@ -98,6 +100,8 @@ pub struct RequestContext {
     access_stats_behavior: AccessStatsBehavior,
     page_content_kind: PageContentKind,
     pub micros_spent_throttled: optional_counter::MicroSecondsCounterU32,
+    /// Total number of kilobytes of layer files processed in this request.
+    pub vectored_access_delta_file_size_kb: AtomicUsize,
```
Comment (arpad-m) [inline pageserver/src/context.rs:104]: Is this used anywhere?

**Review by arpad-m — CHANGES_REQUESTED** [review]
What happens on a pageserver restart? I don't see any persistence of this value, so it's lost whenever we restart the pageserver, until the next basebackup happens.

We should add persistence somehow.

We already have a non-persisted value via #6736 but it's only of limited use as we regularly restart pageservers.

File: pageserver/src/aux_file.rs (hunk 1 above)
Comment (arpad-m) [reply pageserver/src/aux_file.rs:149]: Good point, it's fine.

File: pageserver/src/context.rs (hunk 2 above)
Comment (skyzh) [reply pageserver/src/context.rs:104]: Oh it’s from my other pull request, will remove it.

**Comment by skyzh** [comment]
I don’t plan to persist it because it’s an estimator. If we configure an alert for aux file size and some tenant exceeds the limit, it will fire before the page server restart so it does not matter whether it’s persisted or not?

**Comment by skyzh** [comment]
... or we can have logical size calculation to recompute aux file size in later patches

File: pageserver/src/context.rs (hunk 2 above)
Comment (skyzh) [reply pageserver/src/context.rs]: fixed in https://github.com/neondatabase/neon/pull/7623/commits/13977eae249f82ec38cdaaf2573adfabac971003

**Comment by skyzh** [comment]
Per discussion, we can merge this pull request first and we will have at least some metrics, and add this to initial logical size calculation in the future.

**Pull request merged**